### PR TITLE
Fix process product code and tests for netfx

### LINKED ISSUE #18262: Tests under "System.Diagnostics.Tests.ProcessStartInfoTests" failed
Detail: https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fdesktop~2Fcli~2F/build/20170412.01/workItem/System.Diagnostics.Process.Tests/wilogs

Result:
~~~
2017-04-12 01:49:31,403: INFO: proc(54): run_and_log_output: Output:   Discovering: System.Diagnostics.Process.Tests
2017-04-12 01:49:31,653: INFO: proc(54): run_and_log_output: Output:   Discovered:  System.Diagnostics.Process.Tests
2017-04-12 01:49:31,653: INFO: proc(54): run_and_log_output: Output:   Starting:    System.Diagnostics.Process.Tests
2017-04-12 01:49:32,355: INFO: proc(54): run_and_log_output: Output:     System.Diagnostics.Tests.ProcessStartInfoTests.TestUseShellExecuteProperty_SetAndGet_Windows [FAIL]
2017-04-12 01:49:32,355: INFO: proc(54): run_and_log_output: Output:       Assert.False() Failure
2017-04-12 01:49:32,355: INFO: proc(54): run_and_log_output: Output:       Expected: False
2017-04-12 01:49:32,355: INFO: proc(54): run_and_log_output: Output:       Actual:   True
2017-04-12 01:49:32,355: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2017-04-12 01:49:32,355: INFO: proc(54): run_and_log_output: Output:         E:\A\_work\60\s\corefx\src\System.Diagnostics.Process\tests\ProcessStartInfoTests.cs(243,0): at System.Diagnostics.Tests.ProcessStartInfoTests.TestUseShellExecuteProperty_SetAndGet_Windows()
2017-04-12 01:49:32,403: INFO: proc(54): run_and_log_output: Output:     System.Diagnostics.Tests.ProcessStartInfoTests.TestEnvironmentVariables_Environment_DataRoundTrips [FAIL]
2017-04-12 01:49:32,403: INFO: proc(54): run_and_log_output: Output:       Assert.Throws() Failure
2017-04-12 01:49:32,403: INFO: proc(54): run_and_log_output: Output:       Expected: typeof(System.ArgumentException)
2017-04-12 01:49:32,403: INFO: proc(54): run_and_log_output: Output:       Actual:   (No exception was thrown)
2017-04-12 01:49:32,403: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2017-04-12 01:49:32,403: INFO: proc(54): run_and_log_output: Output:         E:\A\_work\60\s\corefx\src\System.Diagnostics.Process\tests\ProcessStartInfoTests.cs(460,0): at System.Diagnostics.Tests.ProcessStartInfoTests.TestEnvironmentVariables_Environment_DataRoundTrips()
2017-04-12 01:49:32,418: INFO: proc(54): run_and_log_output: Output:     System.Diagnostics.Tests.ProcessStartInfoTests.TestEnvironmentProperty [FAIL]
2017-04-12 01:49:32,418: INFO: proc(54): run_and_log_output: Output:       Assert.Throws() Failure
2017-04-12 01:49:32,418: INFO: proc(54): run_and_log_output: Output:       Expected: typeof(System.ArgumentException)
2017-04-12 01:49:32,418: INFO: proc(54): run_and_log_output: Output:       Actual:   (No exception was thrown)
2017-04-12 01:49:32,418: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2017-04-12 01:49:32,418: INFO: proc(54): run_and_log_output: Output:         E:\A\_work\60\s\corefx\src\System.Diagnostics.Process\tests\ProcessStartInfoTests.cs(48,0): at System.Diagnostics.Tests.ProcessStartInfoTests.TestEnvironmentProperty()
2017-04-12 01:49:34,918: INFO: proc(54): run_and_log_output: Output: System.IndexOutOfRangeException: Index was outside the bounds of the array.
2017-04-12 01:49:34,918: INFO: proc(54): run_and_log_output: Output:    at System.Text.StringBuilder.get_Chars(Int32 index)
2017-04-12 01:49:34,918: INFO: proc(54): run_and_log_output: Output:    at System.Diagnostics.AsyncStreamReader.GetLinesFromStringBuilder()
2017-04-12 01:49:34,918: INFO: proc(54): run_and_log_output: Output:    at System.Diagnostics.AsyncStreamReader.ReadBuffer(IAsyncResult ar)
2017-04-12 01:49:34,918: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
2017-04-12 01:49:34,918: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
2017-04-12 01:49:34,918: INFO: proc(54): run_and_log_output: Output:    at System.IO.Stream.ReadWriteTask.System.Threading.Tasks.ITaskCompletionAction.Invoke(Task completingTask)
2017-04-12 01:49:34,918: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task.FinishContinuations()
2017-04-12 01:49:34,918: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task.Finish(Boolean bUserDelegateExecuted)
2017-04-12 01:49:34,918: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
2017-04-12 01:49:34,918: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task.ExecuteEntry(Boolean bPreventDoubleExecution)
2017-04-12 01:49:34,918: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ThreadPoolWorkQueue.Dispatch()
2017-04-12 01:49:34,934: INFO: proc(54): run_and_log_output: Output: Finished running tests.  End time= 1:49:34.93, Exit code = 1
2017-04-12 01:49:34,934: INFO: proc(58): run_and_log_output: Exit Code: 1
2017-04-12 01:49:34,934: ERROR: scriptrunner(82): _main: Error: No exception thrown, but XUnit results not created
2017-04-12 01:49:34,934: ERROR: helix_test_execution(83): report_error: Error running xunit None
~~~

### LINKED ISSUE #18470: Desktop: System.Diagnostics.Tests.ProcessModuleTests.TestModulesContainsDotnet failed with "Xunit.Sdk.ContainsException"
Failed test: System.Diagnostics.Tests.ProcessModuleTests.TestModulesContainsDotnet

MESSAGE:
Assert.Contains() Failure
Not found: (filter expression)
In value: <CastIterator>d__94<ProcessModule> [System.Diagnostics.ProcessModule (xunit.console.exe), System.Diagnostics.ProcessModule (ntdll.dll), System.Diagnostics.ProcessModule (MSCOREE.DLL), System.Diagnostics.ProcessModule (KERNEL32.dll), System.Diagnostics.ProcessModule (KERNELBASE.dll), ...]
+++++++++++++++++++

STACK TRACE:
at System.Diagnostics.Tests.ProcessModuleTests.TestModulesContainsDotnet() in D:\j\workspace\outerloop_net---903ddde6\src\System.Diagnostics.Process\tests\ProcessModuleTests.cs:line 35

Detail: https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_netfx_windows_nt_debug/21/testReport/System.Diagnostics.Tests/ProcessModuleTests/TestModulesContainsDotnet/

### LINKED ISSUE #18471: Desktop: System.Diagnostics.Tests.ProcessTests.TestInvalidPriorityClass failed with "Xunit.Sdk.ThrowsException"
Failed test: System.Diagnostics.Tests.ProcessTests.TestInvalidPriorityClass

Configuration: outerloop_netfx_windows_nt_debug

Detail: https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_netfx_windows_nt_debug/21/testReport/System.Diagnostics.Tests/ProcessTests/TestInvalidPriorityClass/

MESSAGE:
Assert.Throws() Failure
Expected: typeof(System.ArgumentException)
Actual: typeof(System.ComponentModel.InvalidEnumArgumentException): The value of argument 'value' (96) is invalid for Enum type 'ProcessPriorityClass'.\r\nParameter name: value
+++++++++++++++++++

STACK TRACE:
at System.Diagnostics.Process.set_PriorityClass(ProcessPriorityClass value) 
at System.Diagnostics.Tests.ProcessTests.<>c__DisplayClass53_0.<TestInvalidPriorityClass>b__0() 
in D:\j\workspace\outerloop_net---903ddde6\src\System.Diagnostics.Process\tests\ProcessTests.cs:line 672

### LINKED ISSUE #18472: Desktop: System.Diagnostics.Tests.ProcessTests.Process_StartWithDuplicatePassword failed with "Xunit.Sdk.ThrowsException"
Failed test: System.Diagnostics.Tests.ProcessTests.Process_StartWithDuplicatePassword

Configuration: outerloop_netfx_windows_nt_debug

Detail: https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_netfx_windows_nt_debug/21/testReport/System.Diagnostics.Tests/ProcessTests/Process_StartWithDuplicatePassword/

Message:
~~~
 Assert.Throws() Failure
        Expected: typeof(System.ArgumentException)
        Actual:   typeof(System.InvalidOperationException): The Process object must have the UseShellExecute property set to false in order to start a process as a user.
~~~

Stack Trace:
~~~
at System.Diagnostics.Process.StartWithShellExecuteEx(ProcessStartInfo startInfo) 
at System.Diagnostics.Tests.ProcessTests.<>c__DisplayClass129_0.<Process_StartWithDuplicatePassword>b__0() 
in D:\j\workspace\outerloop_net---903ddde6\src\System.Diagnostics.Process\tests\ProcessTests.cs:line 1464
~~~

### LINKED ISSUE #18473: Desktop: Tests under "System.Diagnostics.Tests.ProcessTests" failed with "Xunit.Sdk.EqualException"
Configuration: outerloop_netfx_windows_nt_debug

~~~
System.Diagnostics.Tests.ProcessTests.TestProcessName [FAIL]
Assert.Equal() Failure
        Expected: RemoteExecutorConsoleApp
        Actual:   xunit.console
        Stack Trace:
          at System.Diagnostics.Tests.ProcessTests.TestProcessName()
in D:\j\workspace\outerloop_net---903ddde6\src\System.Diagnostics.Process\tests\ProcessTests.cs(685,0)

System.Diagnostics.Tests.ProcessTests.TestMainModuleOnNonOSX [FAIL]
        Assert.Equal() Failure
                   (pos 0)
        Expected: dotnet.exe
        Actual:   xunit.console.exe
                   (pos 0)
        Stack Trace:
         at System.Diagnostics.Tests.ProcessTests.TestMainModuleOnNonOSX()
in  D:\j\workspace\outerloop_net---903ddde6\src\System.Diagnostics.Process\tests\ProcessTests.cs(271,0)

System.Diagnostics.Tests.ProcessTests.TestStartInfo [FAIL]
Assert.Equal() Failure
                   (pos 0)
        Expected: D:\\j\\workspace\\outerloop_net---903ddde6\\b···
        Actual:   RemoteExecutorConsoleApp.exe
                   (pos 0)
        Stack Trace:
         at System.Diagnostics.Tests.ProcessTests.TestStartInfo()
in  D:\j\workspace\outerloop_net---903ddde6\src\System.Diagnostics.Process\tests\ProcessTests.cs(924,0)
~~~

Detail: https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_netfx_windows_nt_debug/21/consoleText

### PR DESCRIPTION
**Pull request by hughbe**
Fixes #18262
Fixes #18470
Fixes #18471
Fixes #18472
Fixes #18473

@safern

### CODE REVIEW

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs [hunk 1]
```diff
@@ -43,8 +43,19 @@ public void TestEnvironmentProperty()
...
             Assert.Equal(CountItems + 1, environment.Count);
 
-            //Exception not thrown with invalid key
-            Assert.Throws<ArgumentException>(() => { environment.Add("NewKey2", "NewValue2"); });
+            // Exception thrown with invalid key
+            // The full .NET framework represents Environment with StringDictionary. On .NET Core we use
+            // use Dictionary<string, string>, because we need to have case sensitive behaviour on Unix.
+            // The full .NET Framework uses CaseSensitiveStringDictionary, but we can't have an
+            // equivilent as we don't have access to the internal state of StringDictionary in .NET Core.
+            if (PlatformDetection.IsFullFramework)
```
Comment (hughbe) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: I'm not really sure how to fix this in the product code - the full .NET framework uses CaseSensitiveStringDictionary, which is an internal type that accesses the internal Hashtable in StringDictionary.

**Comment by safern** [comment]
cc: @Priya91

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: I'm investigating Hashtable with some kind of case insensitive comparer

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs [hunk 2]
```diff
@@ -128,8 +139,18 @@ public void TestEnvironmentProperty()
             //Exception not thrown with invalid key
             Assert.Throws<ArgumentNullException>(() => environment.Add(null, "NewValue2"));
 
-            //Invalid Key to add
-            Assert.Throws<ArgumentException>(() => environment.Add("NewKey2", "NewValue2"));
+            // Invalid Key to add
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: "Exception not thrown with invalid key" above should say "Exception thrown with invalid key"? although hardly worth a comment

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs [hunk 3]
```diff
@@ -43,8 +43,19 @@ public void TestEnvironmentProperty()
             environment.Remove("NewKey");
             Assert.Equal(CountItems + 1, environment.Count);
 
-            //Exception not thrown with invalid key
-            Assert.Throws<ArgumentException>(() => { environment.Add("NewKey2", "NewValue2"); });
+            // Exception thrown with invalid key
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: `CountItems` seems like it should be `countItems`

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Hashtable doesn't seem to work very well, we would have to create an adaptor that exactly matches IDictionary -> IDictionary<string, string>.
Short of moving StringDictionary to S.P.CoreLib and then using the internal CaseInsensitiveStringDictionary I don't think there's much we can do.

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs [hunk 4]
```diff
@@ -43,8 +43,19 @@ public void TestEnvironmentProperty()
...
+            // use Dictionary<string, string>, because we need to have case sensitive behaviour on Unix.
+            // The full .NET Framework uses CaseSensitiveStringDictionary, but we can't have an
+            // equivilent as we don't have access to the internal state of StringDictionary in .NET Core.
+            if (PlatformDetection.IsFullFramework)
+            {
+                environment.Add("NewKey2", "NewValue2");
+            }
+            else
+            {
+                Assert.Throws<ArgumentException>(null, () => environment.Add("NewKey2", "NewValue2"));
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Oops, that would be a bug that the parameter name on the exceptoin is `null`..

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs (hunk 1 above)
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Typo `equivilent`.

Comment (safern) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: @hughbe I believe this behavior was ported like this by design. I think it makes the implementation easier and better to understand. This was ported in #13203 by myself.

cc: @stephentoub @danmosemsft any thoughts?

I think we should open an issue as netfx-port-consider to port this to behave as core using dictionary instead of hash table.

Comment (safern) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: I remember when porting this that we decided to do it this way as the contents in StringDictionary are internal and we weren't able to override them to use a custom comparer to behave differently according to the OS, plus that Dictionary works better than HashTable and that way EnviromnentVariables could just be a wrapper around Environment which in both desktop and core is an IDictionary.

Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Without debugging, just looking at the StringDictionaryWithComparer code, which is what it uses on Windows, I'm not sure why it's not throwing when you insert the same key a second time. Is it because it's got the same vlaue? I don't see that...

I'm not sure what you mean by 'how to fix this' -- you mean fix core so it doesn't throw if hte value is the same?

Comment (hughbe) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: I'm not sure how breaking this is. I can imagine a bunch of scenarios where we're adding a bunch of environment variables in, for example from a table/list. If there are duplicates then we'll break a bunch of things.

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs [hunk 5]
```diff
@@ -484,10 +518,21 @@ public void TestEnvironmentVariables_Environment_DataRoundTrips()
 
             string envVarResultKey = "";
             string envVarResultValue = "";
-            foreach(KeyValuePair<string, string> entry in psi.EnvironmentVariables)
+
+            // .NET Core uses a Dictionary<string, string>, but the full .NET Framework uses
+            // StringDictionary. These collections return different types from GetEnumerator.
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Seems like there ought to be a generic way somewhere to compare two IDictionary's.

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: And enumerating through the environment variables is broken too. Desktop code like this doesn't work:
```
foreach (DictionaryEntry entry in psi.EnvironmentVariables)
{

}
```

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs (hunk 4 above)
Comment (hughbe) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Would have to be fixed over in coreclr for Dictionary (https://github.com/dotnet/coreclr/blob/f355c5789d5d7cd6ea10bee326c66f04641cfc90/src/mscorlib/src/System/ThrowHelper.cs#L116)

Would be a good idea, I would imagine it would cause a bunch of param names to have to be nerfed for netfx across corefx tests

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs [hunk 6]
```diff
@@ -496,8 +541,11 @@ public void TestEnvironmentVariables_Environment_DataRoundTrips()
             psi.EnvironmentVariables.Add("NewKey3", "NewValue3");
             KeyValuePair<string, string>[] kvpa = new KeyValuePair<string, string>[5];
             psi.Environment.CopyTo(kvpa, 0);
-            Assert.Equal("NewKey3", kvpa[2].Key);
-            Assert.Equal("NewValue3", kvpa[2].Value);
+
+            // .NET Core uses a Dictionary<string, string>, but the full .NET Framework uses
+            // StringDictionary. These collections order the output of CopyTo differently.
+            Assert.Equal(PlatformDetection.IsFullFramework ? "NewKey2" : "NewKey3", kvpa[2].Key);
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: since the order of hte dictionary is not relevant in the first place (or even well defined) why not just throw in an OrderBy and ther'es no need to check what FX you're running on?

Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: same elsewhere

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs (hunk 4 above)
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Wow, 600 hits in our code for `Assert.Throws<ArgumentException>(null,` ... something for another day. I am pretty confident it's not a prohibited breaking change.

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs (hunk 5 above)
Comment (hughbe) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Unfortunately not, I think:
```cs
Assert.Equal(psi.Environment, psi.EnvironmentVariables);
```

fails to compile:
```cs
Assert.Equal(psi.Environment, (IEnumerable)psi.EnvironmentVariables);
```
fails with NotEqual.

```cs
Assert.Equal(psi.Environment, psi.EnvironmentVariables.Cast<KeyValuePair<string, string>>());
```

works with netcoreapp but fails with an InvalidOperationException with netfx

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs (hunk 1 above)
Comment (safern) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: I see why StringDictionaryWithComparer doesn't throw exceptions when adding a duplicated key. When calling .Add they override it and they call contents[key.ToLower(CultureInfo.CultureInvariant)] and Hashtable indexer property doesn't throws when adding to an existing key obviously it just updates the value inside that key.

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs (hunk 6 above)
Comment (safern) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Good idea.

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs (hunk 5 above)
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: What about some linq?

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs (hunk 1 above)
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: I think we should work to get close to the desktop behavior. As I mentioned earlier I know some code cast to Hashtable before

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs [hunk 7]
```diff
@@ -32,19 +32,29 @@ public void TestEnvironmentProperty()
...
-            Assert.Equal(CountItems + 2, environment.Count);
+            Assert.Equal(countItems + 2, environment.Count);
             environment.Remove("NewKey");
-            Assert.Equal(CountItems + 1, environment.Count);
+            Assert.Equal(countItems + 1, environment.Count);
 
-            //Exception not thrown with invalid key
-            Assert.Throws<ArgumentException>(() => { environment.Add("NewKey2", "NewValue2"); });
+            // The full .NET framework represents Environment with StringDictionary. On .NET Core we use
+            // use Dictionary<string, string>, because we need to have case sensitive behaviour on Unix.
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: If this is the reason, why don't we construct the Dictionary with a case-insensitive comparer on Windows so as to keep the behavior consistent?
cc: @Priya91

Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Actually, we do:
https://source.dot.net/#System.Diagnostics.Process/System/Diagnostics/ProcessStartInfo.cs,78
so I'm not understanding the comment. Can you clarify?

Comment (hughbe) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: The default behaviour of StringDictionary is case insensitive. The full .NET framework uses StringDictionary internally to hold Environment. 
There was no good way to make StringDictionary case insensitive without accessing internal state of StringDictionary, which we can't do in the .NET Core packaged world.

I assumed this is why we went with Dictionary. Maybe I'm wrong!

Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Let me rephrase / change the question:
This looks like an important difference in behavior, and I'm not sure how it relates to casing explanation in the comment. Why are we accepting this behavioral difference rather than fixing it to match?

Comment (hughbe) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: From @safern 

> @hughbe I believe this behavior was ported like this by design. I think it makes the implementation easier and better to understand. This was ported in #13203 by myself.

cc: @stephentoub @danmosemsft any thoughts?

I think we should open an issue as netfx-port-consider to port this to behave as core using dictionary instead of hash table.

Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Just to make sure we're talking about the same thing:
It looks like when you do `processStartInfo.Environment.Add` and the key already exists, in desktop it overwrites the existing key/value whereas currently in core it throws.  That's a big change, introducing an exception where there wasn't one before, and in a scenario that's quite easy to hit.  Why is the core behavior easier/better?  Porting that back to desktop would almost certainly break code and would I expect be a non-starter.

Comment (hughbe) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: I agree, but I'm not sure the best way to fix this without somehow accessing the internal structure of StringDictionary.

Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Why do we need to access StringDictionary internals?  The API returns an `IDictionary<TKey,TValue>`... why can't we return any implementation of that interface we want?

Comment (hughbe) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: The full .NET framework has a couple of differences:
- The `environmentVariables` variable is a StringDictionary (vs Dictionary<string, string in corefx)
- The implementation of `EnvrionmentVariables` creates a dictionary from `Environment`. In corefx this is the other way around. 
- The return value of a StringDictionary in terms of IDictionary<string, string> (https://github.com/Microsoft/referencesource/blob/4fe4349175f4c5091d972a7e56ea12012f1e7170/System/compmod/system/collections/specialized/stringdictionary.cs#L188) has a pretty long and convoluted adapter that maintains the behaviour of StringDictionary for IDictionary<string, string>. This internal state is available to ProcessStartInfo in mscorlib but not in corefx.

If we want to unite with mscorlib behaviour I think we'll need to reimplement the entire IDictionary adapter. It's doable but is a bit of a nuiscance.

Then we'll need to make sure that StringDictionaryWrapper has the correct behaviour that matches with netfx. And we need to use Dictionary<string, string> instead of StringDictionary as we want platform dependant case sensitive that StringDictionary doesn't support.

Comment (safern) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: We could definitely match Desktop behavior with StringDictionaryWrapper by making it implement IDictionary<TKey, TValue> and have its contents to be a Dictionary as it is, and when calling .Add instead of calling contents.Add() just do contents[key] so that we override the value instead of throwing. Being said that we would just need to create the StringDictionaryWrapper with a comparer to be passed to the Dictionary depending on the OS we are running on. 

Then for `Environment` property we could just return `EnvironmentVariables` as it would implement IDictionary.

We would need to use StringDictionary internals if we wanted to use/replace StringDictionary's contents which are a Hashtable and they are overriden in case it needs a comparer, but if we override IDictionary behavior by implementing it we could make it match and use our own contents to be a Dictionary. 

@Hughbe, can you do that?

Comment (Priya91) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: The test here is that on full framework you can add duplicate keys, and the last value wins. Whereas on .NET Core adding duplicate keys throws. That's because we return Dictionary, and Add doesn't allow duplicates. So what @safern suggests sounds good, we can make Environment, return EnvironmentVariables just like on Desktop. But it complicates purpose of adding Environment API, where we wanted to move from non-generic to generic.

Comment (safern) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: We don't have to make `Environment` return `EnvironmentVariables` to match desktop's behavior if we want to conserve generic behavior. We could also make Environment be a wrapper around a Dictionary that implements IDictionary and we could override the behavior of Add to match desktop where we override the keys and then `EnvironmentVariables` could be a wrapper around `Environment` that copies over that wrapper class as its contents and inherits from StringDictionary overriding its behavior to just call the wrapper apis.

Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Also related https://github.com/dotnet/corefx/issues/13146

File: src/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 8]
```diff
@@ -915,44 +915,75 @@ public void TestProcessOnRemoteMachineUnix()
...
+            var process = new Process();
+            Assert.Throws<ArgumentNullException>("value", () => process.StartInfo = null);
+        }
+
+        [Fact]
+        public void StartInfo_GetOnRunningProcess_ThrowsInvalidOperationException()
+        {
+            Process process = Process.GetCurrentProcess();
+
+            // .NET Core fixes a bug where Process.StartInfo for a running process would
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessTests.cs]: "running" might better be worded as "unrelated" ?

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs (hunk 7 above)
Comment (hughbe) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: A wrapper around a wrapper - brilliant. I'll take a look

Comment (safern) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: I know right? That is if we want to preserve generic. If not then we could do my first suggestion. Any final thoughts? @stephentoub @danmosemsft @Priya91

File: src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs [hunk 9]
```diff
@@ -138,17 +128,8 @@ public void TestEnvironmentProperty()
...
-            // the key already exists.
-            if (PlatformDetection.IsFullFramework)
-            {
-                environment.Add("NewKey2", "NewValue2");
-            }
-            else
-            {
-                Assert.Throws<ArgumentException>(null, () => environment.Add("NewKey2", "NewValue2"));
-            }
+            environment.Add("NewKey2", "NewValue2OverridenAgain");
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs]: Nit `overriden` -> `overridden` thoughout

File: src/System.Diagnostics.Process/src/System/Collections/Specialized/DictionaryWrapper.cs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+        public bool IsReadOnly => ((IDictionary)_contents).IsReadOnly;
+        public bool IsSynchronized => ((IDictionary)_contents).IsSynchronized;
+        public bool IsFixedSize => ((IDictionary)_contents).IsFixedSize;
+        public object SyncRoot => ((IDictionary)_contents).SyncRoot;
+
+        public void Add(string key, string value)
+        {
+            // If the key already exists, overwrite it.
+            if (ContainsKey(key))
```
Comment (Priya91) [inline src/System.Diagnostics.Process/src/System/Collections/Specialized/DictionaryWrapper.cs]: No need to check for presence of key, can just do this[key] = value. this indexer set doesn't throw for duplicates, rewrites the value.

**Comment by Priya91** [comment]
@hughbe there;s some test failures on Unix,

```
MESSAGE:
System.InvalidCastException : Unable to cast object of type 'System.Collections.DictionaryEntry' to type 'System.Collections.Generic.KeyValuePair`2[System.String,System.String]'.
+++++++++++++++++++
STACK TRACE:
at System.Diagnostics.Tests.ProcessStartInfoTests.TestEnvironmentVariablesPropertyUnix()
```

**Comment by Priya91** [comment]
@hughbe Any updates on this? Thanks!

File: src/System.Diagnostics.Process/src/System/Collections/Specialized/StringDictionaryWrapper.cs [hunk 11]
```diff
@@ -27,25 +27,38 @@ public StringDictionaryWrapper(Dictionary<string, string> contents)
...
 
-        public override object SyncRoot => ((ICollection)_contents).SyncRoot;
-        
-        public override void Add(string key, string value) => _contents.Add(key, value);
+        public override object SyncRoot => _contents.SyncRoot;
+
+        public override void Add(string key, string value)
+        {
+            if (_contents.ContainsKey(key))
+            {
```
Comment (hughbe) [inline src/System.Diagnostics.Process/src/System/Collections/Specialized/StringDictionaryWrapper.cs]: We do need this one. Without it, I get the following test failures:
```
System.Diagnostics.Tests.ProcessStartInfoTests.TestEnvironmentVariables_Environment_DataRoundTrips [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentException)
        Actual:   (No exception was thrown)
        Stack Trace:
           C:\Users\hughb\Documents\GitHub\corefx\src\System.Diagnostics.Process\tests\ProcessStartInfoTests.cs(471,0):
   at System.Diagnostics.Tests.ProcessStartInfoTests.TestEnvironmentVariables_Environment_DataRoundTrips()
```

**Comment by hughbe** [comment]
Sorry for the delay. I've fixed the unix test failures and addressed your feedback

**Comment by danmosemsft** [comment]
@dotnet-bot test this please (https://github.com/dotnet/corefx/issues/18794)

**Comment by danmosemsft** [comment]
@dotnet-bot test Innerloop Tizen armel Release Cross Build (timeout)

File: src/System.Diagnostics.Process/src/System/Collections/Specialized/StringDictionaryWrapper.cs (hunk 11 above)
Comment (safern) [inline src/System.Diagnostics.Process/src/System/Collections/Specialized/StringDictionaryWrapper.cs]: We do need this as we are calling the DictionaryWrapper.Add and it just overrides the key, on desktop in StringDictionary.Add it calls contents.Add where contents is a Hashtable and Hashtable.Add does throw an ArgumentException when the key is duplicate.
